### DISCUSSION TITLE
Promise-returning function provided to attribute where a void return was expected.

### DISCUSSION THREAD

**Discussion by Adnan0061** [react-hook-form/react-hook-form#8622] [Q&A] [answered]
I am having a typescript-eslint error

checked [this one](https://codesandbox.io/s/react-hook-form-handlesubmit-ts-v7-z9z0g) which is also recommended on [react-hook-form website](https://react-hook-form.com/ts), following the same. 

We are using 
next:12.1.0
react-hook-form:7.24.0

how can I solve it?
Thanks
![Screenshot 2022-07-03 154323](https://user-images.githubusercontent.com/48943431/177034494-e5281158-0a35-4b85-a76e-762867b1624c.jpg)

**Comment by Moshyfawn** [comment]
Try typing your submit handler this way instead:
```ts
import type { SubmitHandler } from "react-hook-form"

const onSubmit: SubmitHandler<FormValues> = (data) => {}
```

**Reply by Adnan0061** [reply]
yes, checked that also and that's why I mentioned it react query doc.
this also doesn't work,

![Screenshot last](https://user-images.githubusercontent.com/48943431/177083816-cffc95e3-fa1f-4d55-b162-feda9ffbebb9.jpg)

**Reply by ItaloCovas** [reply]
You a God, thx hahaha <3

**Comment by gaurangrshah** [comment]
Has anyone solved this? I'm running into the same issue. Is this a typescript version issue? 

```
"react-hook-form": "^7.36.1",
"typescript": "4.7.4"
```

**Reply by Adnan0061** [reply]
this is a tsconfig issue. You can share your ts config, I can probably help you with that.

**Reply by rtman** [reply]
@Adnan0061 if you have fixed this please post your fix.

**Reply by Adnan0061** [reply]
@gaurangrshah @rtman actually this is a showing from eslint configuration. In .eslintrc.json file you might find the following option this is the issue unless "react-hook-form: doesn't change their structure.

"@typescript-eslint/no-misused-promises": "error",

**Reply by rtman** [reply]
yup you will see this if that rule is in use, doesn't mean it's wrong though.

**Comment by gaurangrshah** [comment]
@Adnan0061 @rtman 
I found a work around, well actually two, since hook-form is not the only place I am seeing the error.

The first is to wrap the `handleSubmit` caller in an `onPromise` handler:
![image](https://user-images.githubusercontent.com/1086269/196705458-cf73a914-8300-4b2f-a30b-0cef6e195d1b.png)


The next is to use the void syntax:
![image](https://user-images.githubusercontent.com/1086269/196705764-1b7672d5-fcbf-4712-903b-96e64b4b0d7f.png)

**Reply by gaurangrshah** [reply]
F.Y.I. this is the onPromise helper:

```ts
/**
 *
 *
 * @export
 * @template T
 * @param {(event: React.SyntheticEvent) => Promise<T>} promise
 * @return {*}
 */
export function onPromise<T>(
  // used to wrap react-hook-forms's submit handler
  // https://github.com/react-hook-form/react-hook-form/discussions/8020#discussioncomment-3429261
  promise: (event: React.SyntheticEvent) => Promise<T>
) {
  return (event: React.SyntheticEvent) => {
    if (promise) {
      promise(event).catch((error) => {
        console.error('Unexpected error', error);
      });
    }
  };
}
```

**Reply by Adnan0061** [reply]
great!!!
but why void??

on you should create a pr with this onPromise helper as default inside handleSubmit function.

**Reply by gaurangrshah** [reply]
I'm fairly new to typescript, but from what I can gather void before the function seems to indicate that the function explicitly has no return.

https://www.tutorialsteacher.com/typescript/typescript-void

**Reply by Adnan0061** [reply]
no, I think this is not right. 
()=>void  
this is type declaration for function returning void.
but 
()=>void  signIn('spotify')
I don't think this is a valid code.

**Reply by gaurangrshah** [reply]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void
![image](https://user-images.githubusercontent.com/1086269/197540547-c1c79c3c-6ab2-43ee-8d0d-23d15c134f34.jpeg)

**Reply by Adnan0061** [reply]
my bad, I didn't aware of that. Great!!!

**Reply by gaurangrshah** [reply]
no worries, i wasn't either, but now it's in my arsenal, and I have a better understanding of it. so it's a win / win.

**Reply by peter-novak-m** [reply]
I wasn't aware of this either! Thanks a lot!

**Reply by mkandan** [reply]
Same, thank you

**Comment by Moshyfawn** [comment, marked as answer]
https://github.com/react-hook-form/react-hook-form/discussions/9325#discussioncomment-4060566

> Alright, I've done a little bit more research.
> 
> Check out this typescript-eslint/typescript-eslint#4623 PR that adds granular options within `checksVoidReturns` for [the `no-misused-promises` TS ESLint rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-misused-promises.md); the `attributes` options was added specifically for use-cases like the one you're experiencing. This is the official TS ESLint solution and we recommend you update your `.eslintrc` config file with it:
> 
> ```json
> "rules": {
>   "@typescript-eslint/no-misused-promises": [2, {
>     "checksVoidReturn": {
>       "attributes": false
>     }
>   }]
> },
> ```

**Reply by CesarQD98** [reply]
Interesting info! Just a quick question, what does the `2` stand for?

**Reply by stephenvector** [reply]
@CesarQD98 `2` is the same as `"error"`. It threw me off too, until I found [this doc](https://eslint.org/docs/latest/use/configure/rules#rule-severities).

**Comment by RubenSmn** [comment]
For those who do not want to **disable** the eslint rules. I solved it by passing a anonymous function with the `event` which you can then use to call on `handleSubmit`.

```typescript
<form
  onSubmit={(event) =>
    void handleSubmit(onSubmit)(event)
  }
>
```

**Reply by suchcodemuchwow** [reply]
This one gives:
```ts
Expected 'undefined' and instead saw 'void'.eslint[no-void](https://eslint.org/docs/rules/no-void)
```
So you basically need to disable no-void then 😅

**Reply by dBianchii** [reply]
What are you even doing here? Calling the promise using the event as a parameter?
Can someone explain?
I did this in my code and it is working without changing my eslint rules

**Reply by RubenSmn** [reply]
@dBianchii  the `handleSubmit` method, we get from the `useForm` hook, returns a function that can be called to handle the submit event from the form. We can invoke this function by calling it with the `event` as a argument.

Maybe this way of writing it is more clear:

```typescript
onSubmit={(event) => {
  const theReturnedFunction = handleSubmit(onSubmit);
  void theReturnedFunction(event);
}}
```

**Comment by anricoj1** [comment]
I still think this goes beyond "it's just eslint" or "tsconfig rule". Different ways of writing this causes different outcomes.

I started with this
```tsx
<form onSubmit={form.handleSubmit(onSubmit)}>
```
Which yields this error: "Promise-returning function provided to attribute where a void return was expected". Which makes sense because you're trying to assign a Promise returning function to onSubmit which expects void.

You can do this but this seems to skip any validators you have like a zod schema. It also submits your form as query params.
```tsx
<form onSubmit={(event) => void form.handleSubmit(onSubmit)}>
```

Passing the event to the previous example seems to work as expected which I got from @RubenSmn 's answer.

```tsx
<form onSubmit={(event) => void form.handleSubmit(onSubmit)(event)}>
```

But If you want to go for extra measures you could do it like this as well,
```tsx
 <form onSubmit={(event) => {
      event.preventDefault();
      void form.handleSubmit(onSubmit)(event);
  }}>
```

**Comment by kieracarman** [comment]
It looks like this could be also be written more elegantly like:
```js
<form onSubmit={void handleSubmit(onSubmit)} />
```
Unless I'm missing something?

**Reply by ralph-p** [reply]
This works but it does not over ride the default form behavior with `event.preventDefault();` 
This solution seems to be working but I wonder if there is a better way. 
```
  const onSubmit = (data: FormData) => {
    ...do something with data...
    reset()
  };

 ...

  <form onSubmit={(event) => {
      event.preventDefault();
      void handleSubmit(onSubmit)(event);
  }}>
```

**Comment by Ali7040** [comment]
I am facing the same Promise-returning function provided to attribute where a void return was expected. 

onSelect={async (props) => {
        const linkedEntityId = props.type === EntityType.climb
          ? (props as TypesenseDocumentType).climbUUID
          : (props as TypesenseAreaType).id
        void onAdd({
          mediaId: mediaWithTags.id,
          entityId: linkedEntityId,
          entityType: props.type === EntityType.climb ? TagTargetType.climb : TagTargetType.area
        })
      }  how can I solve

**Comment by tparvi** [comment]
We have solved this with a helper function

```
function asyncHelper(func: Promise<void>) {
  return () => {
    func.catch((error) => {
      console.error(error);
    });
  };
}
```

Usage

```
async function onCLick(): Promise<void> {
  // your async code
}

<Button onClick={() => asynHelper(onClick())}.../>
```

In our case the actual `onClick` handler either just calls something like `trigger()` / react query's `invalidateQueries` so there isn't really anything we can do if those fail. In some cases we might have own try/catch inside the `onClick` handler.

**Reply by buraczyn** [reply]
The problem here is when you want to try/catch the function that returns the promise. This limits that ability to handle it in different ways
